Add a OperatorHandle argument to boxed kernels

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30370 [wip] Use codegen'ed unboxing wrappers
* #29934 Boxed variable dispatch
* #29682 backend fallback test
* #29681 better boxing
* #29337 Hide the OperatorKernel* argument from the stack based kernel API
* #29330 OperatorHandle::callBoxed/callUnboxed
* #29214 Test cases for backend fallback kernels
* **#29201 Add a OperatorHandle argument to boxed kernels**
* #30340 Make Dispatcher::backendFallbackKernels_ an array
* #30333 [pytorch][PR] Remove LeftRight from OperatorEntry and DispatchTable.
* #30332 [pytorch][PR] Convert KernelTable to a flat-indexed array rather than a hashtable.

This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

smessmer committed:
Add a OperatorHandle argument to boxed kernels

This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

ezyang reviewed (approved):
Have we audited the OperatorHandle API so that it is OK to make public?

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer commented:
Which part of the API exactly? This is the current API, but I don't think it's stable yet. The `callUnboxed` method for example has quite some quirks - you need to pass it exactly the same type the kernel function had (i.e. `const Tensor&` is **not** the same as `Tensor`). Ideally, you wouldn't have to specify the type arguments at all and it would infer it from the values you pass in.

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]

smessmer committed:
Update on "Add a OperatorHandle argument to boxed kernels"


This is required for boxed backend fallback kernels (e.g. lazy, AMP) because they need to know which op was actually called.

Differential Revision: [D18282746](https://our.internmc.facebook.com/intern/diff/D18282746/)

[ghstack-poisoned]
